### ISSUE TITLE
More server-client mod synchronization options

### ISSUE THREAD

**Issue by kittenchilly** [tModLoader/tModLoader#1445]
### Description
There should be more server-to-client mod synchronization options than the one we currently have, which is synching normal-sided mods only.
Ideally, there should be three options that the server host should be able to choose when starting up the server:

- Client must strictly match all mods of the server
(You can't have any mods different than the server, including client-side mods. Think of it as some sort of full anti-cheat if you're really picky about that)
- Clients must have all mods of the server at minimum
(You have to have all the mods of the server, including client sided mods, but you can download your own client sided mods that the server doesn't have if you want)
- Clients require all non-client mods at minimum
(The way it currently works.)

Again, this should be chosen by the server host when starting up a server, like how you choose enabled mods and what-not.

### What does this proposal attempt to solve or improve?
If you're a server host, you won't believe how annoying it is that client-side mods aren't synced from server to client. You have to manually tell players to download themselves, which could be preventing them from having optimal gameplay and advantages.
Also, the strict option can prevent people from cheating if a server host really cares about that.

### Which (other) solutions should be considered?
Not any that I can see, other than replacing the current method with the 2nd option in the list instead of adding three options.

**Comment by sgkoishi** [comment]
It's a nice to have feature, but I would say the first one is not that helpful actually - you can't stop a cheating mod, or a modded tModLoader from hiding itself.

The sync-or-not is controlled by the mod itself currently, and adding a new "The server asks you to download a client mod" requires a bit of work (probably a new UI or config file for this).
I actually have a hacky idea for this: create a simple empty synced mod and add the client mods you want as dependencies. When players download this empty mod, they will be prompted to download corresponding client mod. Not sure if this work :)

**Comment by Chicken-Bones** [comment]
A good proposal. One alternative would be to have a config file where the server can list the 'extra' client side mods to sync. Keep in mind that client mods aren't actually loaded on the server. Though the enable/disable menu does still show them.

**Comment by DarioDaF** [comment]
> It's a nice to have feature, but I would say the first one is not that helpful actually - you can't stop a cheating mod, or a modded tModLoader from hiding itself.
> 
> The sync-or-not is controlled by the mod itself currently, and adding a new "The server asks you to download a client mod" requires a bit of work (probably a new UI or config file for this).
> I actually have a hacky idea for this: create a simple empty synced mod and add the client mods you want as dependencies. When players download this empty mod, they will be prompted to download corresponding client mod. Not sure if this work :)

A skeleton of "empty" mod that will ask you if you want to install the others might be a really useful thing to have so server owners can set it up just by listing the optional dependency in a file and pushing Build.

**Comment by kpodjuk** [comment]
This feature would be greatly appreciated. I host a couple servers with different mods on each, server side mod synch works great of course, press a connect button and your done. I love that. But lack of client side synch means that when we play together we aren't really "on the same page" which is important for me. Try to get your friend grup to all install the same mods on calamity playthrough and later disable them for vanilla, no chance. Server config flag to force client mods would be great.

**Comment by JavidPack** [comment]
![image](https://github.com/tModLoader/tModLoader/assets/4522492/ee31e04e-916a-451e-acdf-d097df5c8c9d)

Once #3822 is merged, I think it would be easy to implement servers suggesting clientside mods as another option in this UI. The server could report clientside mods that were in enabled.json but weren't loaded to the client, and the user would be able to check or uncheck mods to download and enable. 

It might be annoying, however, to have to go through this UI each time you connect to a specific server and intentionally have one of the suggested mods disabled, so more thought might be needed on this idea.

**Comment by Chicken-Bones** [comment]
> I agree with the assertion that the first idea isn't helpful. Strictly matching mods is absolutely and utterly useless. There're useful client-side-only mods that only do stuff like change the UI (see: Boss Checklist). Limiting or restricting that entirely isn't very user-friendly, IMO. It's super easy to patch that stuff out with both a pre-modified client or patching at runtime.

The 'strictly match' option could be used to disable a previously distributed client mod. When a server operators who uses tML as the distribution system for their friends decides to remove a client only mod from the playthrough. 

I don't know how likely/useful this would be. 

Unfortunately, adding new features for server operators suffers a massive visibility issue, due to the lack of a 'server control panel'. There's also Host & Play to consider.
